File: 3180-maximum-total-reward-using-operations-i/3180-maximum-total-reward-using-operations-i.rb
# @param {Integer[]} reward_values
# @return {Integer}
def max_total_reward(reward_values)
  reward_values.sort!
  maximum = reward_values[-1]
  dp = Array.new(maximum * 2, 0)

  reward_values.each do |rew|
    (0...rew).each do |i|
      if dp[i + rew] < dp[i] + rew
        dp[i + rew] = dp[i] + rew
      end
    end
  end

  dp.max
end

File: 3180-maximum-total-reward-using-operations-i/3180-maximum-total-reward-using-operations-i.js
/**
 * @param {number[]} rewardValues
 * @return {number}
 */
var maxTotalReward = function(rewardValues) {
  rewardValues.sort((a, b) => a - b);
  let maximum = rewardValues[rewardValues.length - 1];

  //let dp = new Array(maximum * 2).fill(0); //below might be faster/efficient
  let dp = new Int8Array(maximum * 2)

  for (let rew of rewardValues) {
    for (let i = 0; i < rew; i++) {
      if (dp[i + rew] < dp[i] + rew) {
        dp[i + rew] = dp[i] + rew;
      }
    }
  }

  return Math.max(...dp);
};